### PR TITLE
[WIP] Keep connectionstring safe

### PR DESCRIPTION
**Pull request by melkio**
Connects to https://github.com/Particular/PlatformDevelopment/issues/718
(implements https://github.com/Particular/docs.particular.net/issues/1268)

### CODE REVIEW

File: nservicebus/azure/intro-keep-connection-string-safe.include.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+It is possible to accidentally leak sensitive information in the connection string if it's not properly secured. E.g. the information can be leaked if an error occurs when communicating across untrusted boundaries, or if the error information is logged to an unsecured log file.
+
+In order to prevent it, transport, versions 7 and above, allows for creating a logical name for each connection string. The name is mapped to the physical connection string, and connection strings are always referred to by their logical name.
```
Comment (SeanFeldman) [inline nservicebus/azure/intro-keep-connection-string-safe.include.md]: If the message is on the queue, what would `ReplyTo` header contain for example? Would it help to show the difference?

File: nservicebus/azure/intro-keep-connection-string-safe.include.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+It is possible to accidentally leak sensitive information in the connection string if it's not properly secured. E.g. the information can be leaked if an error occurs when communicating across untrusted boundaries, or if the error information is logged to an unsecured log file.
```
Comment (SeanFeldman) [inline nservicebus/azure/intro-keep-connection-string-safe.include.md]: After reading this I'm still not sure how this affects me as an NSB user. Does it mean the log files are not sure? What are the boundaries that the reference made to?

**Comment by SeanFeldman** [comment]
@melkio :+1: Upgrade guide would definitely need to have clear explanation what and why.

File: nservicebus/azure/intro-keep-connection-string-safe.include.md (hunk 2 above)
Comment (melkio) [inline nservicebus/azure/intro-keep-connection-string-safe.include.md]: @SeanFeldman I only brought ASQ related section into an include file and, as suggested by @yvesgoeleven, linked it from a common place to avoid copy&paste and docu duplication.
I did nothing on this paragraphs, I'll work on it soon. I've read your comments about it somewhere ;)

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+- Cloud
+- Azure
+- Transports 
+- Security
+reviewed: 2016-04-26
+---
+
+include: keep-connection-string-safe
+
+To enable this behavior, apply the following configuration:
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: Putting my non-maintainer :tophat: on. 

>To enable this behavior, apply the following configuration:

What is this behaviour? What does it do? Why would I wan to use it?
This is kind of information I'd like to find in the first paragraph and after that see _how_ to do it :)

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: More info:

As a reader, I can be either a newbie, or someone who's already using nsb.asb and wants to read about this new features.

- For a new user there's no existing context (the fact that connection strings where not safe).
- For an existing user they know what's happening, but they don't know about how safe connection strings work 
  - link to the upgrade guide
- For both type of users, how to configure safe connections

Would this flow make sense?

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+After enabling this behavior, if `ReplyTo` message header contains the value `queueName@connectionString`, it will be replaced by `queueName@namespaceName`.
+
+Using `AddNamespace(string name, string connectionString)` configuration API is possible to register namespaces mapping logical names with connection strings, as shown below for `SingleNamespacePartitioning` strategy:
+
+snippet: map_logical_name_to_connection_string 
+
+For detailed explanation about all supported partitioning strategies and how to configure namespaces mapping for them see related [article](multiple-namespaces-support.md).
+
+## How to upgrade endpoints
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: Should this be in the upgrade guide?

File: nservicebus/upgrades/asb-6to7.md [hunk 5]
```diff
@@ -59,4 +59,14 @@ snippet:topology-selection-upgrade-guide
 
 When the `EndpointOrientedTopology` is selected, it is also necessary to configure [publisher names](/nservicebus/azure-service-bus/publisher-names-configuration.md), in order to ensure that subscribers receive event messages:
 
-snippet:publisher_names_mapping_upgrade_guide
\ No newline at end of file
+snippet:publisher_names_mapping_upgrade_guide
+
+## Keep connection string safe
+
+In version 7 and above, to avoid sharing sensitive data contained into connection strings, it's possibile to enable `UseNamespaceNameInsteadOfConnectionString` feature:
```
Comment (SeanFeldman) [inline nservicebus/upgrades/asb-6to7.md]: I'm confused a little. Upgrade guide is intended for thins that have changed between old and new version and **must** be done in order to work properly. If by default, connection strings are used, that means during upgrade there's nothing  one must do in order for solution to work. If enabling safe connection strings is optional, it shouldn't be in the upgrade guide. Please correct me if I'm wrong with assuming that by default we're backwards compatible and one has to opt in to enable safe connection strings. Thank you.

File: nservicebus/azure/keep-connection-string-safe.include.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Transport, versions 6 and below, uses raw connection string provided into message headers. It could result in a leak of sensitive information, f.i. logging it.
```
Comment (SeanFeldman) [inline nservicebus/azure/keep-connection-string-safe.include.md]: Where's this include is used?

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md (hunk 3 above)
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: Overlooked the `keep-connection-string-safe` part.
Would it make sense to elaborate more and provide a little example (tiny snippet of an output) to demonstrate the problem?

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: With snippet like that each time we add something, we have to keep in mind that it's going to be used in multiple places. Is that kind of re-use is desired? I don't think this "duplication" is bad. Plus, the ability of showing what would be rendered in version 6 w/o safe connection strings would be hurt since for ASB and ASQ it's not the same connection string type. My 2c :)

Comment (petersgiles) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: I agree with all of @SeanFeldman's points

**Comment by petersgiles** [comment]
Overall I think  there needs to be a better description of what the feature is, who would use it and when and how.

The snippets seem good.

I'm also a little hazy on what I am required to upgrade and what I don't have to worry about

**Comment by melkio** [comment]
> What is this behaviour? What does it do? Why would I wan to use it?
This is kind of information I'd like to find in the first paragraph and after that see how to do it :)

@SeanFeldman the first paragraph is into the _include_ file. The "behavior" is mentioned there.

**Comment by melkio** [comment]
@Particular/azure-service-bus-maintainers @petersgiles new version. I tried to address your comments. Let me know...

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+component: ASB
+tags:
+- Cloud
+- Azure
+- Transports 
+- Security
+reviewed: 2016-04-26
+---
+
+Transport, versions 6 and below, uses raw connection string provided into message headers.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> Versions 6 and below uses raw the connection string in message headers. For instance, a typical value for `ReplyTo` header could be:

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+- Security
+reviewed: 2016-04-26
+---
+
+Transport, versions 6 and below, uses raw connection string provided into message headers.  
+For instance, a typical value for `ReplyTo` header could be:  
+
+`[queue name]@Endpoint=sb://[namespace name].servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[secrets here]`
+  
+It could result in a leak of sensitive information, f.i. logging it.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: `f.i.`?

Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> This could potentially result in a leak of sensitive information, for example in [log files](/nservicebus/logging/) or in the [error queue](/nservicebus/errors/).

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+reviewed: 2016-04-26
+---
+
+Transport, versions 6 and below, uses raw connection string provided into message headers.  
+For instance, a typical value for `ReplyTo` header could be:  
+
+`[queue name]@Endpoint=sb://[namespace name].servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[secrets here]`
+  
+It could result in a leak of sensitive information, f.i. logging it.  
+In order to prevent it, transport, versions 7 and above, introduces the needs to map logical namespace name to namespace connection string. This configuration allows transport to convert connection string value to namespace name, if mapping exists, for each incoming message, using the logical name internally.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> In order to prevent this Versions 7 and above includes the ability to map logical namespace name to namespace connection string.

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+If a valid mapping between connection string and namespace name doesn't exist, transport throws a `KeyNotFoundException`.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> If a valid mapping between connection string and namespace name doesn't exist a [`KeyNotFoundException`](https://msdn.microsoft.com/en-us/library/system.collections.generic.keynotfoundexception.aspx) is thrown.

Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: also why KeyNotFoundException? should we give them an `Exception` with a sensible message including the key that was attempted?

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+To configure mapping between a namespace name and the corresponding connection string is possible through `AddNamespace(string name, string connectionString)` configuration API, as shown below for `SingleNamespacePartitioning` strategy:
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> Mapping between a namespace name and the corresponding connection string is done as follows:

mentioning inline code when a snippet is below is kind of redundant

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+snippet: map_logical_name_to_connection_string
+
+For detailed explanation about all supported partitioning strategies and how to configure namespaces mapping for them see related [article](multiple-namespaces-support.md).
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: how about 

> For detailed explanation about all supported partitioning strategies, and how to configure namespaces mapping, see [Multiple Namespaces Support](multiple-namespaces-support.md).

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+To configure mapping between a namespace name and the corresponding connection string is possible through `AddNamespace(string name, string connectionString)` configuration API, as shown below for `SingleNamespacePartitioning` strategy:
+
+snippet: map_logical_name_to_connection_string
+
+For detailed explanation about all supported partitioning strategies and how to configure namespaces mapping for them see related [article](multiple-namespaces-support.md).  
+Using directly `ConnectionString(string connectionString)` configuration API, as shown below, transport adds a mapping between a namespace name `default` and the provided connection string.
+
+snippet: map_default_logical_name_to_connection_string 
+
+In the event of an error or when logging only, the logical name will be used avoiding sharing of sensitive information.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: > In the event of an error or when logging only

this is ambiguous. "event of an error" sounds like u are referring to when an exception is thorn.

"when logging only" why "only"?

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+To enable the same behavior also for outgoing messages, a new feature, provided by transport version 7 and above, has to be switched on:
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: How about

> To enable the same behavior also for outgoing messages, a new feature, provided in Version 7 and above, has to be enabled:

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+snippet: enable_use_namespace_name_instead_of_connection_string
+
+Without enabling this behavior, transport converts back the value to `queueName@connectionString` before delivering the message, to ensure backward compatibility among endpoints of different versions.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: i think this would be better done with two headings "versions 6 and below" and "versions 7 and above" and explain the behavior in each of those

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+In the event of an error or when logging only, the logical name will be used avoiding sharing of sensitive information.
+
+To enable the same behavior also for outgoing messages, a new feature, provided by transport version 7 and above, has to be switched on:
+
+snippet: enable_use_namespace_name_instead_of_connection_string
+
+Without enabling this behavior, transport converts back the value to `queueName@connectionString` before delivering the message, to ensure backward compatibility among endpoints of different versions.  
+Enabling `UseNamespaceNameInsteadOfConnectionString` feature, transport ensures the same behavior, avoiding sharing of sensitive information, for all incoming and outgoing messages.
+
+For explanation about how to upgrade endpoints to enable this feature, see [upgrade guide article](/nservicebus/upgrades/asb-6to7.md)
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: make this a related in the header instead 

```
related:
 - nservicebus/upgrades/asb-6to7
```

File: nservicebus/upgrades/asb-6to7.md [hunk 17]
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
...
+snippet:publisher_names_mapping_upgrade_guide
+
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
+In order to be able to enable `UseNamespaceNameInsteadOfConnectionString` follow the next steps:
+- upgrade all endpoints to version 7 or above.
+- switch on `UseNamespaceNameInsteadOfConnectionString` feature one endpoint at time. 
+
+Refer to the [Keep connection string safe page](/nservicebus/azure-service-bus/keep-connectionstring-safe.md) for more details.
```
Comment (SimonCropp) [inline nservicebus/upgrades/asb-6to7.md]: delete this and make the heading a link instead

File: nservicebus/upgrades/asb-6to7.md [hunk 18]
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
...
 When the `EndpointOrientedTopology` is selected, it is also necessary to configure [publisher names](/nservicebus/azure-service-bus/publisher-names-configuration.md), in order to ensure that subscribers receive event messages:
 
-snippet:publisher_names_mapping_upgrade_guide
\ No newline at end of file
+snippet:publisher_names_mapping_upgrade_guide
+
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
+In order to be able to enable `UseNamespaceNameInsteadOfConnectionString` follow the next steps:
```
Comment (SimonCropp) [inline nservicebus/upgrades/asb-6to7.md]: > In order to be able to enable

i think you need to re-word that

File: nservicebus/upgrades/asb-6to7.md [hunk 19]
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
 
 When the `EndpointOrientedTopology` is selected, it is also necessary to configure [publisher names](/nservicebus/azure-service-bus/publisher-names-configuration.md), in order to ensure that subscribers receive event messages:
 
-snippet:publisher_names_mapping_upgrade_guide
\ No newline at end of file
+snippet:publisher_names_mapping_upgrade_guide
+
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).
```
Comment (SimonCropp) [inline nservicebus/upgrades/asb-6to7.md]: what does this have to do with "Keep connection string safe" ?

File: nservicebus/upgrades/asb-6to7.md [hunk 20]
```diff
@@ -59,4 +59,15 @@ snippet:topology-selection-upgrade-guide
...
-snippet:publisher_names_mapping_upgrade_guide
\ No newline at end of file
+snippet:publisher_names_mapping_upgrade_guide
+
+## Keep connection string safe
+
+Transport guarantees backward compatibility between older transport versions (below to version 7) and newer versions (version 7 and above).  
+In order to be able to enable `UseNamespaceNameInsteadOfConnectionString` follow the next steps:
+- upgrade all endpoints to version 7 or above.
+- switch on `UseNamespaceNameInsteadOfConnectionString` feature one endpoint at time.
```
Comment (SimonCropp) [inline nservicebus/upgrades/asb-6to7.md]: why "one endpoint at time"?

Comment (SimonCropp) [inline nservicebus/upgrades/asb-6to7.md]: do people need to deploy all endpoints with v7 prior to enabling UseNamespaceNameInsteadOfConnectionString? if so that is not clear in the above

File: nservicebus/azure-service-bus/keep-connectionstring-safe.md [hunk 21]
```diff
@@ -0,0 +1,40 @@
+---
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/keep-connectionstring-safe.md]: should this be in the menu?

File: nservicebus/upgrades/asb-6to7.md (hunk 20 above)
Comment (melkio) [inline nservicebus/upgrades/asb-6to7.md]: @SimonCropp without deploying v7 transport hasn't `UseNamespaceNameInsteadOfConnectionString` option. Until you have a mixed configuration (v6 and v7 endpoints) this option can't be enabled. it's not backward compatible

I try to rephrase

**Comment by melkio** [comment]
@SimonCropp I addressed all your comments (thanks) except for this one:

> i think this would be better done with two headings "versions 6 and below" and "versions 7 and above" and explain the behavior in each of those

I don't get your point. Could you explain better? Because I don't know how to explain a behavior that doesn't exist on v6 and below :) The first paragraph explains how transport v6 works

File: nservicebus/upgrades/asb-6to7.md (hunk 20 above)
Comment (SeanFeldman) [inline nservicebus/upgrades/asb-6to7.md]: > I'll try to rephrase

:heart:

**Comment by SimonCropp** [comment]
i updated some minor things

**Comment by SimonCropp** [comment]
i think prior to merging this we need to address this https://github.com/Particular/NServiceBus.AzureServiceBus/issues/217

**Comment by SeanFeldman** [comment]
Question: is `asb-credential-warning.include.md` used elsewhere?

**Comment by melkio** [comment]
@SimonCropp take a look to comments here: https://github.com/Particular/NServiceBus.AzureServiceBus/issues/217

**Comment by SimonCropp** [comment]
merged manually here https://github.com/Particular/docs.particular.net/commit/9eb42314b19bb55fed835f9cde2b1bbe4b95abe7
